### PR TITLE
add support of refit-decay

### PR DESCRIPTION
**Pull request by guolinke**

### CODE REVIEW

File: tests/python_package_test/test_engine.py [hunk 1]
```diff
@@ -633,3 +633,25 @@ def is_correctly_constrained(learner):
...
+        evals_result = {}
+        gbm = lgb.train(params, lgb_train,
+                        num_boost_round=20,
+                        valid_sets=lgb_eval,
+                        verbose_eval=False,
+                        evals_result=evals_result)
+        err_pred = log_loss(y_test, gbm.predict(X_test))
+        new_gbm = gbm.refit(X_test, y_test)
+        new_err_pred = log_loss(y_test, new_gbm.predict(X_test))
+        assert err_pred > new_err_pred
```
Comment (StrikerRUS) [inline tests/python_package_test/test_engine.py]: `self.assertGreater`

File: include/LightGBM/config.h [hunk 2]
```diff
@@ -368,6 +368,13 @@ struct Config {
   // desc = see `this file <https://github.com/Microsoft/LightGBM/tree/master/examples/binary_classification/forced_splits.json>`__ as an example
   std::string forcedsplits_filename = "";
 
+  // type = double
+  // default = 0.9
+  // check = >=0.0
+  // check = <=1.0
+  // desc = Decay rate of "refit" task, will use ``leaf_output = refit_decay_rate * old_leaf_output + (1.0 - refit_decay_rate) * new_leaf_output`` to refit trees.
+  double refit_decay_rate = 0.9;
```
[the remark this answers is not in the document]
Comment (guolinke) [reply include/LightGBM/config.h]: refit is supported in python-package after this PR.

[the remark this answers is not in the document]
Comment (StrikerRUS) [reply include/LightGBM/config.h]: Sorry, but I'm a little bit confused. I interpret these parameters like the following: I can add any of them into `params` argument of `train` or `cv` function.
But `refit_decay_rate` parameter can not be used as stand alone param (or I'm mistaken), only as an **argument of `refit` method of Booster**.

BTW, the same about the `task` parameter (I can't use, let say, `params={'task': 'refit', 'refit_decay_rate': 0.5, 'verbose': -1}`).

Maybe for the `task` param it's better to paraphrase like

```
**Note**: can be used only in CLI version; for language-specific packages you can use the correspondent functions
```

And for `refit_decay_rate`: 

```
used only in ``refit`` task in CLI version or as argument in ``refit`` function in language-specific package
```

File: tests/python_package_test/test_engine.py [hunk 3]
```diff
@@ -633,3 +633,24 @@ def is_correctly_constrained(learner):
...
+    def test_refit(self):
+        X, y = load_breast_cancer(True)
+        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
+        params = {
+            'objective': 'binary',
+            'metric': 'binary_logloss',
+            'verbose': -1
+        }
+        lgb_train = lgb.Dataset(X_train, y_train)
+        lgb_eval = lgb.Dataset(X_test, y_test, reference=lgb_train)
```
Comment (StrikerRUS) [inline tests/python_package_test/test_engine.py]: Do we need validation dataset in this test?

File: tests/python_package_test/test_engine.py [hunk 4]
```diff
@@ -633,3 +633,25 @@ def is_correctly_constrained(learner):
...
+        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
+        X_train, X_valid, y_train, y_valid = train_test_split(X_train, y_train, test_size=0.3, random_state=24)
+        params = {
+            'objective': 'binary',
+            'metric': 'binary_logloss',
+            'verbose': -1,
+            'min_data': 10
+        }
+        lgb_train = lgb.Dataset(X_train, y_train)
+        lgb_eval = lgb.Dataset(X_valid, y_valid, reference=lgb_train)
```
Comment (StrikerRUS) [inline tests/python_package_test/test_engine.py]: I probably was unclear, sorry. I meant, `lgb_eval` are not used in this test (neither for early stopping, nor metric evaluation).

Comment (guolinke) [reply tests/python_package_test/test_engine.py]: oh, got you.

**Comment by guolinke** [comment]
@StrikerRUS you are right, I will fix the doc.

File: tests/python_package_test/test_engine.py [hunk 5]
```diff
@@ -633,3 +633,22 @@ def is_correctly_constrained(learner):
...
+        X, y = load_breast_cancer(True)
+        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
+        params = {
+            'objective': 'binary',
+            'metric': 'binary_logloss',
+            'verbose': -1,
+            'min_data': 10
+        }
+        lgb_train = lgb.Dataset(X_train, y_train)
+        evals_result = {}
```
Comment (StrikerRUS) [inline tests/python_package_test/test_engine.py]: `evals_result` is not needed in this test too.

**Comment by StrikerRUS** [comment]
@guolinke I'm afraid that `refit` Python method can suffer from the same [issue](https://github.com/Microsoft/LightGBM/issues/1534) as `predict`, because there are no additional params for `refit`.

**Comment by guolinke** [comment]
@StrikerRUS oh, you are right, we can add kwargs to it.
